pkg/ingest/aggregator.go

```diff
@@ -0,0 +1,320 @@
+	sampleStart time.Time
+	sampleEnd   time.Time
+	minTime     time.Time
+	maxTime     time.Time
+	count       uint64
+	min         float64
+	max         float64
+	sum         float64
+}
+
+type aggregator struct {
+	Resolution   time.Duration
+	aggrsOptions AggrsOptions
+	activeSeries map[uint64]*aggregatedSeries
+	df           AggrDf
+}
+
+// aggregator ingests the data and produces aggregations at the specified resolution.
+type Aggregator struct{ aggregator }
+
+func NewAggregator(resolution time.Duration, optFuncs ...AggrOptionFunc) *aggregator {
+	aggrsOptions := *evalOptions(optFuncs)
+	return &aggregator{
+		Resolution:   resolution,
+		aggrsOptions: aggrsOptions,
+		activeSeries: make(map[uint64]*aggregatedSeries),
+		df:           newAggrDf(aggrsOptions),
+	}
+}
+
+func newAggrDf(ao AggrsOptions) AggrDf {
+	return AggrDf{seriesRecordSets: make(map[uint64]*SeriesRecordSet)}
+}
+
+// Ingest a single chunk provided via an iterator. For a specific searies, We
+// assume the iterator returns values ordered by the timestamp
+// The iterator is expected to already be at the point of the first sample after as.sampleStart
+func (a *aggregator) ingestChunk(as *aggregatedSeries, i input.ChunkIterator) (*aggregatedSeries, error) {
+	var (
+		ts int64
+		v  float64
+		t  time.Time
+	)
+	for {
+		ts, v = i.At()
+		t = timestamp.Time(ts)
+		if t.Before(as.sampleStart) {
+			return as, errors.Errorf("Chunk timestamp %s is less than the sampleStart %s", t, as.sampleStart)
+		}
+		if t.After(as.sampleEnd) {
+			as = a.finalizeSample(as, t)
+		}
+
+		if as.count == 0 {
+			as.minTime = t
+			as.maxTime = t
+			as.min = v
+			as.max = v
+		}
+		if as.maxTime.After(t) {
+			return as, errors.Errorf("Incoming chunks are not sorted by timestamp: expected %s after %s", t, as.maxTime)
+		}
+		as.maxTime = t
+		as.count += 1
+		as.sum += v
+		if as.max < v {
+			as.max = v
+		}
+		if as.min > v {
+			as.min = v
+		}
+		if !i.Next() {
+			break
+		}
+	}
+	return as, nil
+}
+
+func (a *aggregator) addSeriesToDf(as *aggregatedSeries) {
+	rs, ok := a.df.seriesRecordSets[as.hash]
+	if !ok {
+		rs = a.df.addRecordSet(as.labels)
+	}
+
+	vals := map[string]interface{}{
+		"_sample_start": as.sampleStart,
+		"_sample_end":   as.sampleEnd,
+		"_min_time":     as.minTime,
+		"_max_time":     as.maxTime,
+	}
+
+	for _, l := range as.labels {
+		if l.Name == "__name__" {
+			continue
+		}
+		vals[l.Name] = l.Value
+	}
+
+	if a.aggrsOptions.Count.Enabled {
+		vals[a.aggrsOptions.Count.Column] = as.count
+	}
+	if a.aggrsOptions.Sum.Enabled {
+		vals[a.aggrsOptions.Sum.Column] = as.sum
+	}
+	if a.aggrsOptions.Min.Enabled {
+		vals[a.aggrsOptions.Min.Column] = as.min
+	}
+	if a.aggrsOptions.Max.Enabled {
+		vals[a.aggrsOptions.Max.Column] = as.max
+	}
+	rs.Records = append(rs.Records, Record{Values: vals})
+}
+
+// Add the active aggregated series into the final dataframe when we've reached the
+// sample end time. Returns pointer to a new instance of the aggregatedSeries
```

Comment:
```suggestion
// series end time. (?) Returns pointer to a new instance of the aggregatedSeries
```

Reply:
No, this is mean for specific sample: we divide the whole series into samples (each having start and end time), and once we reach to end of one sample, we call this function.